lib/plugins/wallet/cardano/api.js

```diff
@@ -0,0 +1,223 @@
+    })
+      .then(transactions =>
+        transactions
+          .map(item => {
+            const sign = { incoming: +1, outgoing: -1 }[item.direction]
+            return {
+              isConfirmed: item.confirmations >= minConfirmations,
+              amount: sign * item.amount
+            }
+          })
+          .reduce(
+            (acc, item) => {
+              acc.confirmed += item.isConfirmed ? item.amount : 0
+              acc.total += item.amount
+              return acc
+            },
+            {
+              confirmed: 0,
+              total: 0
+            }
+          )
+      )
+      .then(({ confirmed, total }) => ({
+        confirmed: BN(confirmed).round(),
+        total: BN(total).round()
+      }))
+  }
+
+  sumAmountByAddress(movements, address) {
+    return movements
+      .filter(movement => movement.address === address)
+      .map(movement => movement.amount)
+      .reduce((acc, amount) => acc + amount, 0)
+  }
+
+  async addressBalance(
+    address,
+    minConfirmations,
+    mode = CardanoAPI.INCLUDE_ALL_TRANSACTIONS
+  ) {
+    return this.fetchAllPages('transactions', {
+      address: address,
+      per_page: CardanoAPI.PER_PAGE_MAXIMUM
+    })
+      .then(transactions =>
+        transactions
+          .map(transaction => {
+            let amount = 0
+
+            if (mode & CardanoAPI.INCLUDE_INCOME_TRANSACTIONS) {
+              amount -= this.sumAmountByAddress(transaction.inputs, address)
+            }
+
+            if (mode & CardanoAPI.INCLUDE_OUTCOME_TRANSACTIONS) {
+              amount += this.sumAmountByAddress(transaction.output, address)
+            }
+
+            return {
+              isConfirmed: transaction.confirmations >= minConfirmations,
+              amount
+            }
+          })
+          .reduce(
+            (acc, item) => {
+              acc.confirmed += item.isConfirmed ? item.amount : 0
+              acc.total += item.amount
+              return acc
+            },
+            {
+              confirmed: 0,
+              total: 0
+            }
+          )
+      )
+      .then(({ confirmed, total }) => ({
+        confirmed: BN(confirmed)
+          .shift(unitScale)
+          .round(), // TODO: is that shift really need?
+        total: BN(total)
+          .shift(unitScale)
+          .round()
+      }))
+  }
+
+  async sendCoins(walletId, accountIndex, address, cryptoAtoms) {
+    return this.fetch('transactions', null, {
+      destinations: [
+        {
+          amount: cryptoAtoms.toNumber(),
+          address
+        }
+      ],
+      source: {
+        accountIndex,
+        walletId
+      },
+      groupingPolicy: 'OptimizeForHighThroughput'
+    }).then(data => {
+      const inputs = data.inputs.reduce((acc, item) => acc + item.amount, 0)
+      const outputs = data.outputs.reduce((acc, item) => acc + item.amount, 0)
+
+      return {
+        fee: BN(inputs - outputs)
+          .abs()
```

Comment:
huh? why is there `abs` ? In fact, we should probably assert that `inputs >= outputs`

Reply:
yes, this assertion is right, removed